### ISSUE TITLE
Have glib_object_wrapper! generate a get_class() method

### ISSUE THREAD

**Issue by federicomenaquintero** [gtk-rs/glib#243]
From [here][old-ref] - @sdroege suggests that it would be nice to have glib_object_wrapper!() generate 

```
#[repr(C)]
pub struct FooClass(ffi::BarFooClass);

impl IsA<ParentOfFooClass> for FooClass { }
impl IsA<ParentOfParentParentOfFooClass> for FooClass { }

impl Foo {
    pub fn get_class(&self) -> &FooClass {
        unsafe {
            let stash = self.to_glib_none();
            let ptr: ffi::Foo = stash.0;
            let class = *(ptr as *const *const ffi::BarFooClass);
            &*(class as *const FooClass)
        }
}
```

To generate the `FooClass` identifier, we either need the concat_idents!() macro, which is a nightly-only feature, or a procedural macro inside glib-rs that generates such an identifier.

[old-ref]: https://github.com/gtk-rs/glib/pull/234#issuecomment-331371815

**Comment by federicomenaquintero** [comment]
@sdroege - do you mean this?

```
#[repr(C)]
pub struct FooClass(ffi::FooClass); // the ffi class passed to glib_wrapper ?

// IsA<the parent class>?
impl IsA<ParentOfFooClass> for FooClass { }

// etc. for all the parent classes?
impl IsA<ParentOfParentParentOfFooClass> for FooClass { }

impl Foo {
    pub fn get_class(&self) -> &FooClass {
        unsafe {
            let stash = self.to_glib_none();
            let ptr: ffi::Foo = stash.0;
            let class = *(ptr as *const *const ffi::BarFooClass); // This is the **GTypeClass, correct?
            &*(class as *const FooClass)
        }
}
```

**Comment by sdroege** [comment]
Yes, thanks!

**Comment by federicomenaquintero** [comment]
Note to self: once this is done, have gnome-class actually pass in the `imp::FooClass` to glib_wrapper:

```
glib_wrapper ! { 
    pub struct Dummy (Object<imp::Dummy, imp::DummyClass> ) ; 
    ...                                  ^ here
}
```

**Comment by federicomenaquintero** [comment]
Do we also need to `IsA<AllTheInterfaces>`?

**Comment by sdroege** [comment]
Yes

**Comment by sdroege** [comment]
I'm working on this btw

**Comment by federicomenaquintero** [comment]
Is this still relevant?

**Comment by sdroege** [comment]
No, I meant to close this one but forgot. @GuillaumeGomez please do so :)

There are lots of other things going to happen soonish around the macro and the `IsA` trait though but that's unrelated.

There are getters for the class struct now and casts in the class hierarchy. What has to be generated still (coming soon) is wrappers for the class structs for all types.

**Comment by GuillaumeGomez** [comment]
Yes sir!
